// ethr-status-registry/lib/ExternalSignerProvider.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExternalSignerProvider = void 0;
// import { ethers } from 'ethers'
var abstract_signer_1 = require("@ethersproject/abstract-signer");
var properties_1 = require("@ethersproject/properties");
var transactions_1 = require("@ethersproject/transactions");
var ExternalSignerProvider = /** @class */ (function (_super) {
    __extends(ExternalSignerProvider, _super);
    function ExternalSignerProvider(ethSign, provider) {
        var _this = _super.call(this) || this;
        _this.signRaw = function (rawTx) {
            return new Promise(function (resolve, reject) {
                _this.ethSign(rawTx, function (err, signedTx) {
                    if (err) {
                        reject(err);
                    }
                    else {
                        resolve(signedTx);
                    }
                });
            });
        };
        _this.ethSign = ethSign;
        properties_1.defineReadOnly(_this, 'provider', provider);
        return _this;
    }
    ExternalSignerProvider.prototype.connect = function (provider) {
        return new ExternalSignerProvider(this.ethSign, provider);
    };
    ExternalSignerProvider.prototype.sign = function (transaction) {
        return __awaiter(this, void 0, void 0, function () {
            var tx, signed;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, properties_1.resolveProperties(transaction)];
                    case 1:
                        tx = _a.sent();
                        return [4 /*yield*/, this.signRaw(tx)];
                    case 2:
                        signed = _a.sent();
                        return [2 /*return*/, signed];
                }
            });
        });
    };
    ExternalSignerProvider.prototype.getAddress = function () {
        return __awaiter(this, void 0, void 0, function () {
            var signedTx, decoded;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.signRaw({})];
                    case 1:
                        signedTx = _a.sent();
                        decoded = transactions_1.parse(signedTx);
                        return [2 /*return*/, decoded.from];
                }
            });
        });
    };
    ExternalSignerProvider.prototype.signMessage = function (message) {
        return Promise.reject(new Error('Method not implemented.'));
    };
    ExternalSignerProvider.prototype.signTransaction = function (transaction) {
        return Promise.reject(new Error('Method not implemented.'));
    };
    ExternalSignerProvider.prototype.sendTransaction = function (transaction) {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!this.provider) {
                            throw new Error('missing provider');
                        }
                        if (!(transaction.nonce == null)) return [3 /*break*/, 2];
                        _a = transaction;
                        return [4 /*yield*/, this.provider.getTransactionCount(this.getAddress(), 'pending')];
                    case 1:
                        _a.nonce = _b.sent();
                        _b.label = 2;
                    case 2: return [2 /*return*/, this.sign(transaction).then(function (signedTransaction) {
                            return _this.provider.sendTransaction(signedTransaction);
                        })];
                }
            });
        });
    };
    return ExternalSignerProvider;
}(abstract_signer_1.Signer));
exports.ExternalSignerProvider = ExternalSignerProvider;
